### ISSUE TITLE
Any way to avoid full-text index requirement at build-time? (E.g. for using LocalDB even though some queries use full-text indexes)

### ISSUE THREAD

**Issue by cmeeren** [fsprojects/FSharp.Data.SqlClient#384]
I have for a while now used SqlClient with several queries that use full-text indexes. This is causing some major headaches on the build server, so I'm wondering if there is (or could be easy to implement) a way to sidestep the problem entirely.

I have a DB project in Visual Studio to manage the DB schema, consisting mostly of table definitions and sprocs. Since I need full-text indexes, I can't use LocalDB, and I must instead use SQL Server Developer Edition locally. The compile-time connection strings used by SqlClient reflect this. Whenever I make changes to the schema, I deploy to my local instance and reload the project that contains the SqlClient code.

Locally, this works fine. It's a one-time setup of SQL Server on my dev machine, and then everything works.

However, the Azure Pipelines build agents have only LocalDB pre-installed, so I can't deploy the DB project directly. In order for the project to build, the build script must first manually install SQL Server *for each build*, which has the following significant downsides:
* The build frequently fails because the SQL Server install aborts if it detects a pending system reboot, and the Azure Pipelines VMs frequently have pending reboots. In the worst case, I'm pressing "retry", waiting a minute until build fails, press "retry" again, and so on for half an hour until I finally get an agent without a pending reboot and the build completes.
* Each build takes much longer, e.g. 15 minutes instead of 2 minutes, because installing SQL Server takes a long time.

This is *almost* enough to make me reconsider using type providers, or at least SqlClient.

However, before I give up, I just wanted to ask if there is any way to sidestep this issue. Again, the core of the matter is that I am using SqlClient with sprocs that use full-text indexes. If there is a way to make this work with LocalDB and not require the full SQL Server, that would be awesome and completely solve the problem. Unfortunately I don't know whether this is possible.

One feature that probably would solve the problem entirely, is something similar to [SqlProvider's SaveContextSchema](http://fsprojects.github.io/SQLProvider/core/parameters.html#ContextSchemaPath). This saves a JSON file with a representation of the the DB schema, and AFAIK the file can be checked in and used instead of a connection to an actual DB.

Note that, if it matters, I don't *have* to use sprocs. Currently I use them only because the VS tooling is good, with quick Intellisense and completion of names, etc.

**Comment by cmeeren** [comment]
Do you think something like the aforementioned SaveContextSchema could be implemented in SqlClient? If this is something you would consider accepting, do you have any pointers on where in the code I could start experimenting to implement such a feature? I guess it boils down to serializing and deserializing some internal model of the schema?

**Comment by cmeeren** [comment]
I have now worked around the problem by using self-hosted build agents with SQL server installed, so this is no longer a crucial itch for me to scratch and I need to prioritize my time elsewhere. FWIW I still think it would make a good addition to SqlClient at some point. The ability to remove the build-time DB dependency on CI servers would greatly simplify CI.

**Comment by smoothdeveloper** [comment]
@cmeeren the workaround for now is to have a development database which support the feature.

**Comment by cmeeren** [comment]
> @cmeeren the workaround for now is to have a development database which support the feature.

If you have a DB outside of the build agents, then that could break when running multiple builds in parallel. To be clear, my workaround is to have the DB as part of the build agent and deploy/migrate it there during build.

---

As I said above, I still think that an "offline mode" for this type provider would be very useful. Is that something that could be considered at some point, possibly if someone else wants to work on it? If so, it could be good to keep this issue open, or I can open another issue specifically for that feature request.

**Comment by smoothdeveloper** [comment]
>  I can open another issue specifically for that feature request.

#194 is tracking it.

> Is that something that could be considered at some point

If anyone has the opportunity (in free time or sponsored) to contribute such a PR, yes it will be considered.
